### PR TITLE
add keepalive functionality to booting agents

### PR DESCRIPTION
**Pull request by genisd**
This helps with tcp loadbalancers which would otherwise timeout the
connection. Having the connection timeout is problematic because the
drone-agent does not recognize that it lost the connection.

This keepalive features was added in drone-agent 0.8.4 hence we bump
the default version to 0.8.4

### CODE REVIEW

**Comment by genisd** [comment]
@tboerger ah I didn't know that. Will do thanks!
It's still wip, I'll let you guys know once it's ready

**Comment by genisd** [comment]
Is this the best place to put this? It didn't really `feel` clear to me what the best location was for this.

Also do you know where the docs for are for the autoscaler? If possible I would like to create a PR for this as well

File: config/config.go [hunk 1]
```diff
@@ -35,10 +35,12 @@ type (
...
 		Agent struct {
-			Host        string
-			Token       string
-			Image       string `default:"drone/agent:0.8"`
-			Concurrency int    `default:"2"`
+			Host             string
+			Token            string
+			Image            string `default:"drone/agent:0.8"`
+			Concurrency      int    `default:"2"`
+			KeepaliveTime    string `default:"6m"`
```
Comment (tboerger) [inline config/config.go]: Both types should be time.Duration

Comment (genisd) [reply config/config.go]: Done. I was having a bit of trial and error and some missunderstandings.
I'll try this PR (`f252e2a`) on our environment now

Comment (genisd) [reply config/config.go]: It doesn't build in our environment. I think I had this before, a `git commit --amend` and force push actually yields the previous commit beeing actually built/tested.

I had tried time.Duration before and got the same error infact hence I changed it to strings.
I believe time.Duration if casted to a string get's printed out as an int in nanoseconds (though I surely can fix that in the print statement and print it as say 300s (in seconds basically)).

I'll carry on tomorrow I guess, thanks for the quick feedback through! Appreciated.

The error is as follows: 

```
=== RUN   TestLoad
--- FAIL: TestLoad (0.00s)
	load_test.go:138: json: cannot unmarshal string into Go struct field .KeepaliveTime of type time.Duration
FAIL
```

Comment (genisd) [reply config/config.go]: I changed it to nanoseconds in the load_test. That should build successfully

**Comment by genisd** [comment]
turns out I was wrong.
```
package main

import (
	"fmt"
	"time"
)

func main() {
	keepaliveTime := time.Duration(360000000000)
	keepaliveTimeout := time.Duration(30000000000)
	fmt.Println(fmt.Sprintf("DRONE_KEEPALIVE_TIME=%s", keepaliveTime))
	fmt.Println(fmt.Sprintf("DRONE_KEEPALIVE_TIMEOUT=%s", keepaliveTimeout))
}
```

prints 
```
DRONE_KEEPALIVE_TIME=6m0s
DRONE_KEEPALIVE_TIMEOUT=30s
```

not sure if `6m0s` will work for the `DRONE_KEEPALIVE_TIME`. I'll let you know :-)

**Comment by genisd** [comment]
This has been running well for us.
Only thing is, is it okay to have this enabled by default?
It's not ENV override-able, should I build that?

**Comment by genisd** [comment]
I've set `DRONE_KEEPALIVE_TIME` to 1 minute to reduce the amount of time for it to detect that the connection is dead.

**Comment by genisd** [comment]
don't merge this. with `1m` this is lower than the serverside enforced default of 5m
default *has to be* >=6m to sync up with serverside default
probably has to be configurable too. it's kinda neccessarry

we're still playing around with this a little ourselves on google kubernetes with it's loadbalancer

**Comment by genisd** [comment]
set the default to 6m / 30s again as suggested in the docs (and enforced by the grpc server side)

**Comment by yellowmegaman** [comment]
This totally fixed our issues (running about 10-11 hours)

https://discourse.drone.io/t/autoscaler-gce-drone-agents-require-restart-after-idling/1964

**Comment by genisd** [comment]
this is good to go as far as i'm concerned. the only "flaw" in the code is that one could pass in stuff like `1h1m1s` and the `1h` would not get parsed and get silently ignored. Imho not a very big flaw :-)

File: engine/install.go [hunk 2]
```diff
@@ -191,3 +214,26 @@ func (i *installer) errorUpdate(ctx context.Context, server *autoscaler.Server,
 	}
 	return err
 }
+
+func ParseMinuteSecondTimeDuration(str string) time.Duration {
```
Comment (bradrydzewski) [inline engine/install.go]: fwiw there is a duration parser in the standard library https://godoc.org/time#ParseDuration

but in this case the durations are already being parsed and set in the Config structure, so you should not need to re-parse this information.

File: engine/install.go [hunk 3]
```diff
@@ -123,6 +128,22 @@ poller:
 		Str("image", i.image).
 		Msg("create agent container")
 
+	keepaliveTimeEnv := os.Getenv("DRONE_AGENT_KEEPALIVE_TIME")
+	if keepaliveTimeEnv != "" {
```
Comment (bradrydzewski) [inline engine/install.go]: this data was already parsed when the `Config` structure was loaded by the system, and should not need to be re-parsed. Also setting `i.keepaliveTime` this way can cause a race condition, and could result in the program crashing.

File: config/config.go [hunk 4]
```diff
@@ -35,10 +35,12 @@ type (
...
-			Host        string
-			Token       string
-			Image       string `default:"drone/agent:0.8"`
-			Concurrency int    `default:"2"`
+			Host             string
+			Token            string
+			Image            string        `default:"drone/agent:0.8"`
+			Concurrency      int           `default:"2"`
+			KeepaliveTime    time.Duration `default:"6m"`
+			KeepaliveTimeout time.Duration `default:"30s"`
```
Comment (bradrydzewski) [inline config/config.go]: note that when you declare these variables here, they are automatically being parsed from the environment when the system loads. In this case, the system will look for environment variables using `DRONE_AGENT_KEEPALIVETIME` and `DRONE_AGENT_KEEPLIVETIMEOUT` (no space). Note that you can customize the environment variable names. See some of the other attributes in the Config object for examples or refer to https://github.com/kelseyhightower/envconfig

Comment (genisd) [reply config/config.go]: Thank you for all the feedback.
I was a bit puzzled how certain env vars did get parsed, your comments clear up a lot here.

I hope to have some time Thursday/Friday to apply your feedback.

**Comment by yellowmegaman** [comment]
Still working perfectly BTW 😂

**Comment by tboerger** [comment]
> Still working perfectly BTW 😂

But still missing changes, so it won't get merged like it is.

**Comment by frebib** [comment]
@genisd Any chance on this getting fixed so we can have it merged please?

**Comment by tboerger** [comment]
Could be nice to have that for Drone 0.8, but since Drone 0.9 will drop the GRPC communication I'm pretty sure we won't need that anymore (for Drone >= 0.9).

**Comment by frebib** [comment]
Oh I wasn't aware that 0.9 replaced GRPC (it's kind of hard to follow that without a commit log :roll_eyes:). Currently for us, Autoscaler causes more issues than it solves with the constant disconnecting of agents :/ I hope 0.9 resolves the issues

**Comment by tboerger** [comment]
With 0.9 we will switch back to simple RESTless APIs, so I'm pretty sure the connection will improve because of solid battle-tested infrastructure :)

**Comment by bradrydzewski** [comment]
merged via #27
